indigo_optional_drivers/ccd_rpi/indigo_ccd_rpi.cpp

```diff
@@ -0,0 +1,1749 @@
+
+	static indigo_device focuser_template =
+		INDIGO_DEVICE_INITIALIZER("", focuser_attach, indigo_focuser_enumerate_properties, focuser_change_property, NULL, focuser_detach);
+
+	new_eid = "-1";
+	std::lock_guard<std::mutex> lock(device_mutex);
+	RegisterCameraManager *cameraManager_ = RegisterCameraManager::getInstance();
+
+	for (std::string id : cameraManager_->GetCameraIDs()) {
+
+		string cameraName = GetCameraName(id);
+		if (IsNew(id)) {
+			char name[128] = "RPI Camera ";
+			strcat(name, cameraName.c_str());
+
+			char *end = name + strlen(name) - 1;
+			while (end > name && isspace((unsigned char)*end))
+				end--;
+			end[1] = '\0';
+			rpi_private_data *private_data = (rpi_private_data *)indigo_safe_malloc(sizeof(rpi_private_data));
+			private_data->eid = id;
+
+			indigo_device *device = (indigo_device *)indigo_safe_malloc_copy(sizeof(indigo_device), &ccd_template);
+			indigo_device *master_device = device;
+			device->master_device = master_device;
+			snprintf(device->name, INDIGO_NAME_SIZE, "%s", name);
+			indigo_make_name_unique(device->name, "%s", cameraName.c_str());
+			device->private_data = private_data;
+
+			indigo_device *focuser = (indigo_device *)indigo_safe_malloc_copy(sizeof(indigo_device), &focuser_template);
+			focuser->master_device = device;
+			snprintf(focuser->name, INDIGO_NAME_SIZE, "%s (focuser)", name);
+			indigo_make_name_unique(device->name, "%s", cameraName.c_str());
+			focuser->private_data = private_data;
+			private_data->focuser = focuser;
+
+			for (int j = 0; j < MAX_DEVICES; j++) {
+				if (devices[j] == NULL) {
+					indigo_attach_device(devices[j] = device);
+					break;
+				}
+			}
+		}
+	}
+}
+
+indigo_result indigo_ccd_rpi(indigo_driver_action action, indigo_driver_info *info) {
+	static indigo_driver_action last_action = INDIGO_DRIVER_SHUTDOWN;
+
+	SET_DRIVER_INFO(info, "Raspberry Pi Camera", __FUNCTION__, DRIVER_VERSION, true, last_action);
+
+	if (action == last_action)
+		return INDIGO_OK;
+
+	switch (action) {
+	case INDIGO_DRIVER_INIT:
+		last_action = action;
+		for (int i = 0; i < MAX_DEVICES; i++) {
+			devices[i] = 0;
+		}
+		// CSI cameras use dedicated ports which
+		// is not supported by libusb. So no hot-plug
+		// support. Directly attach the camera divice and
+		// let camera module list the attached sensors
+		attach_rpi_ccd();
+		return INDIGO_OK;
+
+	case INDIGO_DRIVER_SHUTDOWN:
+		for (int i = 0; i < MAX_DEVICES; i++)
+			VERIFY_NOT_CONNECTED(devices[i]);
+		last_action = action;
+
+		for (int i = MAX_DEVICES - 1; i >= 0; i--) {
+			indigo_device *device = devices[i];
+			if (device) {
+				indigo_detach_device(device);
+				if (device->master_device == device) {
+					rpi_private_data *private_data = PRIVATE_DATA;
+					if (private_data->buffer != NULL) {
+						free(private_data->buffer);
+					}
+					if (PRIVATE_DATA->focuser) {
```

Comment:
once you move focuser in the devices you do not need a special handling for it here. please check how it is done in other drivers.